packages/instrumentor/edgeIdStrategy.ts

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright 2023 Code Intelligence GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as lock from "proper-lockfile";
+import * as fs from "fs";
+import * as os from "os";
+import process from "process";
+
+import { fuzzer } from "@jazzer.js/fuzzer";
+
+export interface EdgeIdStrategy {
+	nextEdgeId(): number;
+	startForSourceFile(filename: string): void;
+	commitIdCount(filename: string): void;
+}
+
+export abstract class IncrementingEdgeIdStrategy implements EdgeIdStrategy {
+	protected constructor(protected _nextEdgeId: number) {}
+
+	nextEdgeId(): number {
+		fuzzer.coverageTracker.enlargeCountersBufferIfNeeded(this._nextEdgeId);
+		return this._nextEdgeId++;
+	}
+
+	abstract startForSourceFile(filename: string): void;
+	abstract commitIdCount(filename: string): void;
+}
+
+export class MemorySyncIdStrategy extends IncrementingEdgeIdStrategy {
+	constructor() {
+		super(0);
+	}
+
+	// eslint-disable-next-line @typescript-eslint/no-unused-vars
+	startForSourceFile(filename: string): void {
+		// nothing to do here
+	}
+
+	// eslint-disable-next-line @typescript-eslint/no-unused-vars
+	commitIdCount(filename: string) {
+		// nothing to do here
+	}
+}
+
+interface EdgeIdInfo {
+	filename: string;
+	firstId: number;
+	idCount: number;
+}
+
+/**
+ * A strategy for edge ID generation that synchronizes the IDs assigned to a source file
+ * with other processes via the specified `idSyncFile`. The edge information stored as a
+ * line of the format: <source file path>,<initial edge ID>,<total edge count>
+ *
+ * This class takes care of synchronizing the access to the file between
+ * multiple processes accessing it during instrumentation.
+ */
+export class FileSyncIdStrategy extends IncrementingEdgeIdStrategy {
```

Comment:
We probably should create unit tests for the edge id strategies. If this is time critical, we can do it in a separate PR.

Reply:
I've added a unit test for the sync file ID strategy and refactored the interface if the `instrument` functions to make it easier to use and test. Could you have a look?